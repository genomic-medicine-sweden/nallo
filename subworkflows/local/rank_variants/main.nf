//
// A subworkflow to score and rank variants.
//

include { GENMOD_ANNOTATE  } from '../../../modules/nf-core/genmod/annotate/main'
include { GENMOD_MODELS    } from '../../../modules/nf-core/genmod/models/main'
include { GENMOD_SCORE     } from '../../../modules/nf-core/genmod/score/main'
include { GENMOD_COMPOUND  } from '../../../modules/nf-core/genmod/compound/main'
include { TABIX_BGZIPTABIX } from '../../../modules/nf-core/tabix/bgziptabix/main'

workflow RANK_VARIANTS {

    take:
    ch_vcf                // channel: [mandatory] [ val(meta), path(vcf) ]
    ch_pedfile            // channel: [mandatory] [ val(meta), path(ped) ]
    ch_reduced_penetrance // channel: [mandatory] [ val(meta), path(pentrance) ]
    ch_score_config       // channel: [mandatory] [ val(meta), path(ini) ]

    main:
    ch_versions = Channel.empty()

    GENMOD_ANNOTATE ( ch_vcf )
    ch_versions = ch_versions.mix(GENMOD_ANNOTATE.out.versions)

    GENMOD_MODELS (
        GENMOD_ANNOTATE.out.vcf,
        ch_pedfile.map { meta, ped -> ped },
        ch_reduced_penetrance.map { meta, file -> file }
    )
    ch_versions = ch_versions.mix(GENMOD_MODELS.out.versions)

    GENMOD_SCORE (
        GENMOD_MODELS.out.vcf,
        ch_pedfile.map { meta, ped -> ped },
        ch_score_config.map { meta, file -> file }
    )
    ch_versions = ch_versions.mix(GENMOD_SCORE.out.versions)

    GENMOD_COMPOUND ( GENMOD_SCORE.out.vcf )
    ch_versions = ch_versions.mix(GENMOD_COMPOUND.out.versions)

    TABIX_BGZIPTABIX ( GENMOD_COMPOUND.out.vcf )
    ch_versions = ch_versions.mix(TABIX_BGZIPTABIX.out.versions)

    emit:
    vcf      = TABIX_BGZIPTABIX.out.gz_tbi.map { meta, vcf, tbi -> [ meta, vcf ] } // channel: [ val(meta), path(vcf) ]
    tbi      = TABIX_BGZIPTABIX.out.gz_tbi.map { meta, vcf, tbi -> [ meta, tbi ] } // channel: [ val(meta), path(tbi) ]
    versions = ch_versions                                                         // channel: [ path(versions.yml) ]
}
