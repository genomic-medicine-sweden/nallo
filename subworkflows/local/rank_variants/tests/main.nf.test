nextflow_workflow {

    name "Test Workflow RANK_VARIANTS"
    script "../main.nf"
    config "./nextflow.config"
    workflow "RANK_VARIANTS"

    setup {
        run("PREPARE_GENOME") {
            script "../../prepare_genome.nf"
            process {
                """
                input[0] = Channel.of([
                    [ id:'hg38' ],
                    file(params.pipelines_testdata_base_path + 'reference/hg38.test.fa.gz', checkIfExists: true)
                ])
                input[1] = true
                input[2] = [
                    [ id: 'vep_cache' ],
                    file(params.pipelines_testdata_base_path + 'reference/vep_cache_test_data.tar.gz', checkIfExists:true)
                ]
                input[3] = true
                input[4] = Channel.of([
                    file(params.pipelines_testdata_base_path + 'reference/vep_plugin_files.csv', checkIfExists: true)
                ])
                """
            }
        }
        run("MINIMAP2_ALIGN") {
            script "../../../../modules/nf-core/minimap2/align/main.nf"
            process {
                """
                input[0] = [
                    [ id: 'test', num_intervals:1 ],
                    file(params.pipelines_testdata_base_path + 'testdata/HG002_PacBio_Revio.fastq.gz', checkIfExists: true)
                ]
                input[1] = PREPARE_GENOME.out.fasta
                input[2] = true
                input[3] = 'csi'
                input[4] = false
                input[5] = false
                """
            }
        }
        run("SHORT_VARIANT_CALLING") {
            script "../../short_variant_calling/main.nf"
            process {
                """
                input[0] = MINIMAP2_ALIGN.out.bam
                            .join(MINIMAP2_ALIGN.out.index)
                            .join(Channel.of([
                                [ id: 'test', num_intervals:1 ],
                                file(params.pipelines_testdata_base_path + 'reference/test_data.bed', checkIfExists: true)
                            ]))
                input[1] = PREPARE_GENOME.out.fasta
                input[2] = PREPARE_GENOME.out.fai
                input[3] = [[],[]]
                input[4] = [[],[]]
            """
            }
        }
        run("SNV_ANNOTATION") {
            script "../../snv_annotation/main.nf"
            process {
                """
                input[0] = SHORT_VARIANT_CALLING.out.combined_bcf
                input[1] = [
                    file(params.pipelines_testdata_base_path + 'reference/cadd.v1.6.hg38.test_data.zip', checkIfExists: true)
                ]
                input[2] = PREPARE_GENOME.out.fasta
                input[3] = PREPARE_GENOME.out.fai
                input[4] = PREPARE_GENOME.out.vep_resources
                input[5] = Channel.value('110')
                input[6] = PREPARE_GENOME.out.vep_extra_files
                input[7] = false
                input[8] = Channel.value([])
                input[9] = null
                input[10] = null
                """
            }
        }
        run("ANNOTATE_CSQ_PLI") {
            script "../../../../subworkflows/local/annotate_consequence_pli.nf"
            process {
                """
                input[0] = SNV_ANNOTATION.out.vcf
                input[1] = [
                    [ id: 'variant_consequences' ],
                    [ file(params.pipelines_testdata_base_path + 'reference/variant_consequences_v2.txt', checkIfExists: true) ]
                ]
                """
            }
        }
        run("CREATE_PEDIGREE_FILE") {
            script "../../../../modules/local/create_pedigree_file/main.nf"
            process {
                """
                input[0] = [
                    'project',
                    [
                        [ id: 'test', project: 'project', maternal_id: '0', paternal_id: '0', family_id: 'family', sex: '2', phenotype: '2' ]
                    ]
                ]
                """
            }
        }
    }

    test("1 sample") {

        when {
            workflow {
                """
                input[0] = ANNOTATE_CSQ_PLI.out.vcf
                input[1] = CREATE_PEDIGREE_FILE.out.ped
                input[2] = Channel.of([
                    [ id: 'reduced_penetrance' ],
                    file(params.pipelines_testdata_base_path + 'reference/reduced_penetrance.tsv', checkIfExists: true)
                ])
                input[3] = Channel.of([
                    [ id: 'score_config_snv' ],
                    file(params.pipelines_testdata_base_path + 'reference/rank_model_snv.ini', checkIfExists: true)
                ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert workflow.out.tbi.get(0).get(1).endsWith("tbi") },
                { assert snapshot(
                    path(workflow.out.vcf.get(0).get(1)).vcf.summary, // Can't snapshot MD5 because VEP output is unstable
                ).match() }
            )
        }
    }

    test("1 sample - stub") {

        options "-stub"

        when {
            workflow {
                """
                input[0] = ANNOTATE_CSQ_PLI.out.vcf
                input[1] = CREATE_PEDIGREE_FILE.out.ped
                input[2] = Channel.of([
                    [ id: 'reduced_penetrance' ],
                    file(params.pipelines_testdata_base_path + 'reference/reduced_penetrance.tsv', checkIfExists: true)
                ])
                input[3] = Channel.of([
                    [ id: 'score_config_snv' ],
                    file(params.pipelines_testdata_base_path + 'reference/rank_model_snv.ini', checkIfExists: true)
                ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert workflow.out.tbi.get(0).get(1).endsWith("tbi") },
                { assert snapshot(workflow.out).match() }
            )
        }
    }
}
