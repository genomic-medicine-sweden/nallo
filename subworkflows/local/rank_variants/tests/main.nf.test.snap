{
    "1 sample": {
        "content": [
            "VcfFile [chromosomes=[chrX, chr16], sampleCount=1, variantCount=121, phased=false, phasedAutodetect=false]"
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-18T13:47:43.144635306"
    },
    "1 sample - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_data.bed",
                            "project": null,
                            "contains_affected": false
                        },
                        "test_data.bed.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test_data.bed",
                            "project": null,
                            "contains_affected": false
                        },
                        "test_data.bed.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    "versions.yml:md5,58de7735498cf84a4284050260b6f929",
                    "versions.yml:md5,789b8f834c7851ecd93b3488aa0236da",
                    "versions.yml:md5,a1990357a1ce026de4ba955c877c2430",
                    "versions.yml:md5,f56d4e953de59ea23842d33c16e914cf",
                    "versions.yml:md5,f9b69655d279838966e636810ae15f5a"
                ],
                "tbi": [
                    [
                        {
                            "id": "test_data.bed",
                            "project": null,
                            "contains_affected": false
                        },
                        "test_data.bed.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "vcf": [
                    [
                        {
                            "id": "test_data.bed",
                            "project": null,
                            "contains_affected": false
                        },
                        "test_data.bed.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,58de7735498cf84a4284050260b6f929",
                    "versions.yml:md5,789b8f834c7851ecd93b3488aa0236da",
                    "versions.yml:md5,a1990357a1ce026de4ba955c877c2430",
                    "versions.yml:md5,f56d4e953de59ea23842d33c16e914cf",
                    "versions.yml:md5,f9b69655d279838966e636810ae15f5a"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-18T13:48:10.616656902"
    }
}